File: JavaScript/base64_utils/base64_utils_test.js
/**
 * Base64 Utilities - 单元测试
 * 
 * 运行方式:
 * - Node.js: node base64_utils_test.js
 * - 浏览器: 在控制台加载 mod.js 后运行 testAll()
 */

// 加载模块（Node.js 环境）
const Base64Utils = (typeof require !== 'undefined') 
  ? require('./mod.js') 
  : { encode, decode, toUrlSafe, fromUrlSafe, isValid, toUint8Array, fromUint8Array, encodedLength };

const { encode, decode, toUrlSafe, fromUrlSafe, isValid, toUint8Array, fromUint8Array, encodedLength } = Base64Utils;

// 测试结果统计
let passed = 0;
let failed = 0;

/**
 * 断言函数
 */
function assert(condition, message) {
  if (condition) {
    passed++;
    console.log(`✓ ${message}`);
  } else {
    failed++;
    console.error(`✗ ${message}`);
  }
}

/**
 * 测试 encode 函数
 */
function testEncode() {
  console.log('\n=== Testing encode() ===');
  
  // 基本编码测试
  assert(encode('Hello World') === 'SGVsbG8gV29ybGQ=', 'Basic encoding');
  assert(encode('') === '', 'Empty string encoding');
  assert(encode('A') === 'QQ==', 'Single character encoding');
  assert(encode('AB') === 'QUI=', 'Two characters encoding');
  assert(encode('ABC') === 'QUJD', 'Three characters encoding');
  
  // URL 安全编码
  assert(encode('Hello World', { urlSafe: true, pad: false }) === 'SGVsbG8gV29ybGQ', 'URL-safe encoding without padding');
  assert(encode('>>>???', { urlSafe: true }) === 'Pj4-Pz8_', 'URL-safe encoding with special chars');
  
  // 无填充编码
  assert(encode('A', { pad: false }) === 'QQ', 'Encoding without padding');
  
  // Unicode 测试
  assert(encode('你好') === '5L2g5aW9', 'Chinese encoding');
  assert(encode('🎉') === '8J+OiQ==', 'Emoji encoding');
  assert(encode('日本語') === '5pel5pys6Kqe', 'Japanese encoding');
  
  // 错误处理
  try {
    encode(123);
    assert(false, 'Should throw TypeError for non-string input');
  } catch (e) {
    assert(e instanceof TypeError, 'Throws TypeError for non-string input');
  }
}

/**
 * 测试 decode 函数
 */
function testDecode() {
  console.log('\n=== Testing decode() ===');
  
  // 基本解码测试
  assert(decode('SGVsbG8gV29ybGQ=') === 'Hello World', 'Basic decoding');
  assert(decode('') === '', 'Empty string decoding');
  assert(decode('QQ==') === 'A', 'Single character decoding');
  assert(decode('QUI=') === 'AB', 'Two characters decoding');
  assert(decode('QUJD') === 'ABC', 'Three characters decoding');
  
  // URL 安全解码
  assert(decode('SGVsbG8gV29ybGQ', { urlSafe: true }) === 'Hello World', 'URL-safe decoding');
  assert(decode('Pj4-Pz8_', { urlSafe: true }) === '>>>???', 'URL-safe decoding with special chars');
  
  // Unicode 解码
  assert(decode('5L2g5aW9') === '你好', 'Chinese decoding');
  assert(decode('8J+OiQ==') === '🎉', 'Emoji decoding');
  assert(decode('5pel5pys6Kqe') === '日本語', 'Japanese decoding');
  
  // 错误处理
  try {
    decode(123);
    assert(false, 'Should throw TypeError for non-string input');
  } catch (e) {
    assert(e instanceof TypeError, 'Throws TypeError for non-string input');
  }
  
  try {
    decode('Invalid!!!');
    assert(false, 'Should throw Error for invalid Base64');
  } catch (e) {
    assert(e instanceof Error, 'Throws Error for invalid Base64');
  }
}

/**
 * 测试 toUrlSafe 函数
 */
function testToUrlSafe() {
  console.log('\n=== Testing toUrlSafe() ===');
  
  assert(toUrlSafe('SGVsbG8gV29ybGQ=') === 'SGVsbG8gV29ybGQ', 'Convert to URL-safe (no padding)');
  assert(toUrlSafe('SGVsbG8gV29ybGQ=', true) === 'SGVsbG8gV29ybGQ=', 'Convert to URL-safe (with padding)');
  assert(toUrlSafe('a+b/c=') === 'a-b_c', 'Replace + and /');
  
  // 错误处理
  try {
    toUrlSafe(123);
    assert(false, 'Should throw TypeError for non-string input');
  } catch (e) {
    assert(e instanceof TypeError, 'Throws TypeError for non-string input');
  }
}

/**
 * 测试 fromUrlSafe 函数
 */
function testFromUrlSafe() {
  console.log('\n=== Testing fromUrlSafe() ===');
  
  assert(fromUrlSafe('SGVsbG8gV29ybGQ') === 'SGVsbG8gV29ybGQ=', 'Convert from URL-safe');
  assert(fromUrlSafe('a-b_c') === 'a+b/c===', 'Replace - and _');
  assert(fromUrlSafe('QQ') === 'QQ==', 'Add padding for 2 chars');
  assert(fromUrlSafe('QUI') === 'QUI=', 'Add padding for 3 chars');
  
  // 错误处理
  try {
    fromUrlSafe(123);
    assert(false, 'Should throw TypeError for non-string input');
  } catch (e) {
    assert(e instanceof TypeError, 'Throws TypeError for non-string input');
  }
}

/**
 * 测试 isValid 函数
 */
function testIsValid() {
  console.log('\n=== Testing isValid() ===');
  
  // 有效 Base64
  assert(isValid('SGVsbG8gV29ybGQ=') === true, 'Valid Base64 with padding');
  assert(isValid('SGVsbG8gV29ybGQ') === true, 'Valid Base64 without padding');
  assert(isValid('') === false, 'Empty string is invalid');
  
  // 无效 Base64
  assert(isValid('Invalid!!!') === false, 'Invalid characters');
  assert(isValid('QQQQ') === true, 'Valid length (mod 4 = 0)');
  
  // URL 安全验证
  assert(isValid('SGVsbG8gV29ybGQ', { urlSafe: true }) === true, 'Valid URL-safe Base64');
  assert(isValid('a+b/c=', { urlSafe: true }) === false, 'Standard Base64 is invalid for URL-safe check');
  assert(isValid('a-b_cA', { urlSafe: true }) === true, 'Valid URL-safe with - and _');
  
  // 填充选项
  assert(isValid('SGVsbG8gV29ybGQ=', { allowPadding: false }) === false, 'Padding not allowed');
  assert(isValid('SGVsbG8gV29ybGQ', { allowPadding: false }) === true, 'No padding with allowPadding: false');
}

/**
 * 测试 toUint8Array 函数
 */
function testToUint8Array() {
  console.log('\n=== Testing toUint8Array() ===');
  
  const arr1 = toUint8Array('SGVsbG8gV29ybGQ=');
  assert(arr1 instanceof Uint8Array, 'Returns Uint8Array');
  assert(arr1.length === 11, 'Correct length');
  assert(arr1[0] === 72 && arr1[1] === 101, 'Correct values');
  
  const arr2 = toUint8Array('');
  assert(arr2.length === 0, 'Empty string returns empty array');
  
  // 错误处理
  try {
    toUint8Array(123);
    assert(false, 'Should throw TypeError for non-string input');
  } catch (e) {
    assert(e instanceof TypeError, 'Throws TypeError for non-string input');
  }
}

/**
 * 测试 fromUint8Array 函数
 */
function testFromUint8Array() {
  console.log('\n=== Testing fromUint8Array() ===');
  
  const bytes = new Uint8Array([72, 101, 108, 108, 111]);
  assert(fromUint8Array(bytes) === 'SGVsbG8=', 'Basic encoding from Uint8Array');
  
  const empty = new Uint8Array([]);
  assert(fromUint8Array(empty) === '', 'Empty array returns empty string');
  
  // URL 安全编码
  const special = new Uint8Array([62, 62, 62, 63, 63, 63]);
  assert(fromUint8Array(special, { urlSafe: true }) === 'Pj4-Pz8_', 'URL-safe encoding from Uint8Array');
  
  // 错误处理
  try {
    fromUint8Array('not an array');
    assert(false, 'Should throw TypeError for non-Uint8Array input');
  } catch (e) {
    assert(e instanceof TypeError, 'Throws TypeError for non-Uint8Array input');
  }
}

/**
 * 测试 encodedLength 函数
 */
function testEncodedLength() {
  console.log('\n=== Testing encodedLength() ===');
  
  assert(encodedLength('Hello') === 8, 'Length with padding');
  assert(encodedLength('Hello', { pad: false }) === 7, 'Length without padding');
  assert(encodedLength('A') === 4, 'Single char length');
  assert(encodedLength('AB') === 4, 'Two chars length');
  assert(encodedLength('ABC') === 4, 'Three chars length');
  assert(encodedLength('') === 0, 'Empty string length');
  assert(encodedLength(123) === 0, 'Non-string returns 0');
}

/**
 * 测试编解码一致性
 */
function testRoundTrip() {
  console.log('\n=== Testing Round-trip Consistency ===');
  
  const testCases = [
    'Hello World',
    'The quick brown fox jumps over the lazy dog',
    '你好，世界！',
    '🎉🎊🎁',
    '日本語テキスト',
    'Special chars: !@#$%^&*()',
    '',
    'A',
    'AB',
    'ABC'
  ];
  
  for (const testCase of testCases) {
    const encoded = encode(testCase);
    const decoded = decode(encoded);
    assert(decoded === testCase, `Round-trip: "${testCase.substring(0, 20)}`);
  }
}

/**
 * 运行所有测试
 */
function testAll() {
  console.log('Starting Base64 Utils Tests...\n');
  
  testEncode();
  testDecode();
  testToUrlSafe();
  testFromUrlSafe();
  testIsValid();
  testToUint8Array();
  testFromUint8Array();
  testEncodedLength();
  testRoundTrip();
  
  console.log('\n=== Test Summary ===');
  console.log(`Passed: ${passed}`);
  console.log(`Failed: ${failed}`);
  console.log(`Total: ${passed + failed}`);
  
  if (failed === 0) {
    console.log('\n✓ All tests passed!');
    return true;
  } else {
    console.log(`\n✗ ${failed} test(s) failed`);
    return false;
  }
}

// Node.js 环境自动运行测试
if (typeof module !== 'undefined' && require.main === module) {
  const success = testAll();
  process.exit(success ? 0 : 1);
}

File: Python/base64_utils/base64_utils_test.py
"""
AllToolkit - Python Base64 Utilities Test Suite

Comprehensive test suite for Base64 encoding/decoding utilities.
Run with: python -m pytest base64_utils_test.py -v
"""

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from mod import Base64Utils, encode, decode, encode_urlsafe, decode_urlsafe, is_valid


def test_encode_string():
    """Test encoding strings to Base64."""
    assert Base64Utils.encode("Hello, World!") == "SGVsbG8sIFdvcmxkIQ=="
    assert Base64Utils.encode("") == ""
    assert Base64Utils.encode("a") == "YQ=="
    assert Base64Utils.encode("ab") == "YWI="
    assert Base64Utils.encode("abc") == "YWJj"
    print("✓ test_encode_string passed")


def test_encode_bytes():
    """Test encoding bytes to Base64."""
    assert Base64Utils.encode(b"Hello, World!") == "SGVsbG8sIFdvcmxkIQ=="
    assert Base64Utils.encode(b"\x00\x01\x02\xff") == "AAEC/w=="
    print("✓ test_encode_bytes passed")


def test_decode_string():
    """Test decoding Base64 to strings."""
    assert Base64Utils.decode("SGVsbG8sIFdvcmxkIQ==") == "Hello, World!"
    assert Base64Utils.decode("") == ""
    assert Base64Utils.decode("YQ==") == "a"
    assert Base64Utils.decode("YWI=") == "ab"
    assert Base64Utils.decode("YWJj") == "abc"
    print("✓ test_decode_string passed")


def test_decode_to_bytes():
    """Test decoding Base64 to bytes."""
    assert Base64Utils.decode_to_bytes("AAEC/w==") == b"\x00\x01\x02\xff"
    assert Base64Utils.decode_to_bytes("SGVsbG8=") == b"Hello"
    print("✓ test_decode_to_bytes passed")


def test_encode_urlsafe():
    """Test URL-safe Base64 encoding."""
    # String with characters that differ between standard and URL-safe
    result = Base64Utils.encode_urlsafe("hello+world/test")
    assert "+" not in result
    assert "/" not in result
    assert "-" in result or "_" in result or result == "aGVsbG8rd29ybGQvdGVzdA=="
    print("✓ test_encode_urlsafe passed")


def test_encode_urlsafe_no_padding():
    """Test URL-safe Base64 encoding without padding."""
    result = Base64Utils.encode_urlsafe("hello", padding=False)
    assert "=" not in result
    assert result == "aGVsbG8"
    print("✓ test_encode_urlsafe_no_padding passed")


def test_decode_urlsafe():
    """Test URL-safe Base64 decoding."""
    # aGVsbG8- decodes to bytes: hello + 0xfb
    result1 = Base64Utils.decode_urlsafe("aGVsbG8-")
    assert result1.startswith("hello")
    result2 = Base64Utils.decode_urlsafe("aGVsbG8_")
    assert result2.startswith("hello")
    print("✓ test_decode_urlsafe passed")


def test_decode_urlsafe_no_padding():
    """Test decoding URL-safe Base64 without padding."""
    # "hello" encoded without padding
    assert Base64Utils.decode_urlsafe("aGVsbG8") == "hello"
    print("✓ test_decode_urlsafe_no_padding passed")


def test_to_urlsafe():
    """Test converting standard Base64 to URL-safe."""
    standard = "aGVsbG8+/test="
    urlsafe = Base64Utils.to_urlsafe(standard)
    assert "+" not in urlsafe
    assert "/" not in urlsafe
    assert "-" in urlsafe
    assert "_" in urlsafe
    print("✓ test_to_urlsafe passed")


def test_to_urlsafe_no_padding():
    """Test converting to URL-safe without padding."""
    standard = "aGVsbG8+/test="
    urlsafe = Base64Utils.to_urlsafe(standard, padding=False)
    assert "=" not in urlsafe
    print("✓ test_to_urlsafe_no_padding passed")


def test_from_urlsafe():
    """Test converting URL-safe Base64 to standard."""
    urlsafe = "aGVsbG8-_test"
    standard = Base64Utils.from_urlsafe(urlsafe)
    assert "-" not in standard
    assert "_" not in standard
    assert "+" in standard
    assert "/" in standard
    print("✓ test_from_urlsafe passed")


def test_is_valid_standard():
    """Test validating standard Base64."""
    assert Base64Utils.is_valid("SGVsbG8=") == True
    assert Base64Utils.is_valid("YWJj") == True
    assert Base64Utils.is_valid("") == True
    assert Base64Utils.is_valid("Invalid!") == False
    assert Base64Utils.is_valid("aGVsbG8-") == False  # URL-safe in standard mode
    print("✓ test_is_valid_standard passed")


def test_is_valid_urlsafe():
    """Test validating URL-safe Base64."""
    assert Base64Utils.is_valid("aGVsbG8-", urlsafe=True) == True
    assert Base64Utils.is_valid("aGVsbG8_", urlsafe=True) == True
    assert Base64Utils.is_valid("SGVsbG8=", urlsafe=True) == True  # Standard is also valid URL-safe
    assert Base64Utils.is_valid("Invalid!", urlsafe=True) == False
    print("✓ test_is_valid_urlsafe passed")


def test_is_valid_non_string():
    """Test validation with non-string input."""
    assert Base64Utils.is_valid(123) == False
    assert Base64Utils.is_valid(None) == False
    print("✓ test_is_valid_non_string passed")


def test_encoded_length():
    """Test calculating encoded length."""
    assert Base64Utils.encoded_length(0) == 0
    assert Base64Utils.encoded_length(1) == 4  # 1 byte -> 2 chars + 2 padding
    assert Base64Utils.encoded_length(2) == 4  # 2 bytes -> 3 chars + 1 padding
    assert Base64Utils.encoded_length(3) == 4  # 3 bytes -> 4 chars
    assert Base64Utils.encoded_length(100) == 136
    print("✓ test_encoded_length passed")


def test_encoded_length_no_padding():
    """Test calculating encoded length without padding."""
    assert Base64Utils.encoded_length(1, padding=False) == 2
    assert Base64Utils.encoded_length(2, padding=False) == 3
    assert Base64Utils.encoded_length(3, padding=False) == 4
    print("✓ test_encoded_length_no_padding passed")


def test_decoded_max_length():
    """Test calculating max decoded length."""
    assert Base64Utils.decoded_max_length(0) == 0
    assert Base64Utils.decoded_max_length(4) == 3
    assert Base64Utils.decoded_max_length(8) == 6
    assert Base64Utils.decoded_max_length(136) == 102
    print("✓ test_decoded_max_length passed")


def test_roundtrip():
    """Test encode/decode roundtrip."""
    test_cases = [
        "Hello, World!",
        "The quick brown fox jumps over the lazy dog",
        "1234567890",
        "!@#$%^&*()",
        "Unicode: 你好世界 🌍",
        "",
        "a",
        "ab",
        "abc",
    ]
    
    for original in test_cases:
        encoded = Base64Utils.encode(original)
        decoded = Base64Utils.decode(encoded)
        assert decoded == original, f"Failed for: {original}"
    
    print("✓ test_roundtrip passed")


def test_roundtrip_urlsafe():
    """Test URL-safe encode/decode roundtrip."""
    test_cases = [
        "hello+world/test",
        "user+name@example.com",
        "path/to/file",
        "a+b/c=d",
    ]
    
    for original in test_cases:
        encoded = Base64Utils.encode_urlsafe(original)
        decoded = Base64Utils.decode_urlsafe(encoded)
        assert decoded == original, f"Failed for: {original}"
    
    print("✓ test_roundtrip_urlsafe passed")


def test_roundtrip_bytes():
    """Test bytes encode/decode roundtrip."""
    test_cases = [
        b"\x00\x01\x02\x03",
        b"\xff\xfe\xfd\xfc",
        b"\x00\x00\x00\x00",
        b"\xff\xff\xff\xff",
        bytes(range(256)),
    ]
    
    for original in test_cases:
        encoded = Base64Utils.encode(original)
        decoded = Base64Utils.decode_to_bytes(encoded)
        assert decoded == original, f"Failed for: {original}"
    
    print("✓ test_roundtrip_bytes passed")


def test_convenience_functions():
    """Test convenience module-level functions."""
    assert encode("test") == Base64Utils.encode("test")
    assert decode("dGVzdA==") == Base64Utils.decode("dGVzdA==")
    assert encode_urlsafe("test") == Base64Utils.encode_urlsafe("test")
    assert decode_urlsafe("dGVzdA==") == Base64Utils.decode_urlsafe("dGVzdA==")
    assert is_valid("dGVzdA==") == Base64Utils.is_valid("dGVzdA==")
    print("✓ test_convenience_functions passed")


def test_error_handling():
    """Test error handling for invalid input."""
    try:
        Base64Utils.encode(123)
        assert False, "Should have raised TypeError"
    except TypeError:
        pass
    
    try:
        Base64Utils.decode("Invalid!!!")
        assert False, "Should have raised ValueError"
    except ValueError:
        pass
    
    print("✓ test_error_handling passed")


def run_all_tests():
    """Run all tests."""
    print("=" * 60)
    print("AllToolkit - Python Base64 Utilities Test Suite")
    print("=" * 60)
    
    tests = [
        test_encode_string,
        test_encode_bytes,
        test_decode_string,
        test_decode_to_bytes,
        test_encode_urlsafe,
        test_encode_urlsafe_no_padding,
        test_decode_urlsafe,
        test_decode_urlsafe_no_padding,
        test_to_urlsafe,
        test_to_urlsafe_no_padding,
        test_from_urlsafe,
        test_is_valid_standard,
        test_is_valid_urlsafe,
        test_is_valid_non_string,
        test_encoded_length,
        test_encoded_length_no_padding,
        test_decoded_max_length,
        test_roundtrip,
        test_roundtrip_urlsafe,
        test_roundtrip_bytes,
        test_convenience_functions,
        test_error_handling,
    ]
    
    passed = 0
    failed = 0
    
    for test in tests:
        try:
            test()
            passed += 1
        except Exception as e:
            print(f"✗ {test.__name__} failed: {e}")
            failed += 1
    
    print("=" * 60)
    print(f"Results: {passed} passed, {failed} failed")
    print("=" * 60)
    
    return failed == 0


if __name__ == "__main__":
    success = run_all_tests()
    sys.exit(0 if success else 1)